tools/mserver/mserver5.1.in
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH MSERVER5 1 "APRIL 2011" MonetDB "MonetDB Applications"
.SH NAME
mserver5 \- the MonetDB server version 5
.SH SYNOPSIS
.B mserver5
[
.I options
]
.SH DESCRIPTION
.I Mserver5
is the current MonetDB server that performs all processing on request
of clients for a certain database.
.P
Note that while
.I mserver5
is the process that does the actual work, it is usually more common to
start, monitor and connect to the
.I mserver5
process through
.IR monetdbd (1).
.P
This man-page describes the options that
.I mserver5
understands.
It is intended for people who really need to work with
.I mserver5
itself.
In regular cases, the programs
.IR monetdbd (1)
and
.IR monetdb (1)
control the many options, and allow to adjust them to appropriate
values where sensible.
For normal usage, it is preferred to apply any configuration through
these programs.
.SH OPERATION
When the build-time configuration did not disable this, a
.I mserver5
process presents the user with a console prompt.
On this prompt, MAL commands can be executed.  The architecture is
setup to handle multiple streams of requests.
The first thread started represents the server, which is the console
prompt, reading from standard input and writing to standard output.
.P
The server thread started remains in existence until all other threads
die.
The server is stopped by sending it a termination signal (SIGINT,
SIGTERM, SIGQUIT).
.SH MSERVER5 OPTIONS
.I Mserver5
can be started with options as arguments.
.TP
\fB\-\-dbpath=\fP\fIpath\fP
Path where
.I mserver5
should find a database.
Shorthand for option
.BR gdk_dbpath .
Default value:
.BR @LOCALSTATEDIR@/monetdb5/dbfarm/demo .
.TP
\fB\-\-dbextra=\fP\fIpath\fP
Path where mserver5 should store transient data.  Default value is the
value of the
.B \-\-dbpath
option.
.TP
\fB\-\-config=\fP\fIfile\fP
Config file to read options from.
This file can contain all options as can be set with the --set flag.
See
.IR "CONFIG FILE FORMAT" .
.TP
\fB\-\-set\fP \fIoption\fP\fB=\fP\fIvalue\fP
Set individual configuration option.
For possible options, see
.I PARAMETERS
sections.
.TP
.B \-\-help
Print list of options.
.TP
.B \-\-version
Print version and compile configuration.
.TP
.B \-\-in\-memory
Run mserver5 in-memory.
.B No
data will be persisted.
.TP
.B \-\-trace
Print information about MAL instructions as they are executed.
.TP
\fB\-\-verbose\fP[\fB=\fP\fIvalue\fP]
.PD 0
.TP
\fB\-v\fP[\fIvalue\fP]
Set or increase verbosity level.  If no
.I value
is specified, the verbosity level is increased by
.BR 1 .
If a
.I value
is specified, the verbosity level is set to the given value.
.TP
\fB\-\-debug\fP[\fB=\fP\fIvalue\fP]
.PD 0
.TP
\fB\-d\fP[\fIvalue\fP]
Set debug level.  This is mostly for debugging purposes.
The value is an integer, which can be (a bit-wise OR of):
.EX
          \fB1\fP = THRDMASK     = thread-specific debug output
          \fB2\fP = CHECKMASK    = property enforcing on new BATs
          \fB4\fP = MEMMASK      = memory allocation
          \fB8\fP = PROPMASK     = property checking on all values:
                             tells about wrongly set properties
         \fB16\fP = IOMASK       = major IO activity
         \fB32\fP = BATMASK      = BAT handling
.\"         \fB64\fP = PARSEMASK    = parser debugging
        \fB128\fP = PARMASK      = Thread management
.\"        \fB256\fP = HEADLESSMASK = Warn about BAT heads that are not "headless-ready"
        \fB512\fP = TMMASK       = Transaction management
       \fB1024\fP = TEMMASK      = Locks and Triggers
.\"       \fB2048\fP = DLMASK       = Dynamic loading
       \fB4096\fP = PERFMASK     = BBP Performance (?)
       \fB8192\fP = DELTAMASK    = Delta debugging (?)
      \fB16384\fP = LOADMASK     = Module loading
.\"      \fB32768\fP = YACCMASK     = Yacc specific error messages
    \fB2097152\fP = ALGOMASK     = show join/select algorithm chosen
    \fB4194304\fP = ESTIMASK     = show result size estimations
                             (for join, select)
.\"    \fB8388608\fP = XPROPMASK    = extended property checking:
.\"                             tells also about not set properties
   \fB16777216\fP = NOSYNCMASK   = disable forcefully synchronizing files
                             to disk.  If this flag is set, there is
                             no guarantee that the database remains
                             consistent after a crash.  DO NOT USE
                             (unless you really don't care about your
                             data).
   \fB33554432\fP = DEADBEEFMASK = disable "cleaning" of freed memory
                             in GDKfree() (e.g., for performance
                             measurements)
   \fB67108864\fP = ALLOCMASK    = exhaustive GDK malloc & free tracing
                             for debugging (GDK developers, only)
  \fB134217728\fP = OPTMASK      = trace the actions, decisions and
                             effects of MAL optimizers
  \fB268435456\fP = HEAPMASK     = trace/debug HEAPextend;
                             used only for development & debugging
  \fB536870912\fP = FORCEMITOMASK = forcefully activate mitosis even on
                              small tables, i.e., split small tables
                              in as many (tiny) pieces as there are
                              cores (threads) available;
                              this allows us to test mitosis
                              functionality without requiring large
                              data sets (\(em at the expense of a
                              potentially significant interpretation
                              overhead for unnecessarily large plans);
                              used only for development & testing;
                              set automatically by Mtest.py
.EE
The
.I value
together with the
.B =
sign is optional.  If not specified, it defaults to
.BR 1 .
In the short form \fB\-d\fP, the value, if present, must immediately
(i.e. without space) follow the option.
The values of multiple instances of this flag are OR-ed together.
.TP
.B \-\-threads
Equivalent to \fB\-\-debug=\fP\fI(THRDMASK | PARMASK)\fP.
.TP
.B \-\-memory
Equivalent to \fB\-\-debug=\fP\fI(MEMMASK | ALLOCMASK)\fP.
.TP
.B \-\-properties
Equivalent to \fB\-\-debug=\fP\fI(CHECKMASK | PROPMASK | BATMASK)\fP.
.TP
.B \-\-io
Equivalent to \fB\-\-debug=\fP\fI(IOMASK | PERFMASK)\fP.
.TP
.B \-\-heaps
Equivalent to \fB\-\-debug=\fP\fI(HEAPMASK)\fP.
.TP
.B \-\-transactions
Equivalent to \fB\-\-debug=\fP\fI(TMMASK | DELTAMASK | TEMMASK)\fP.
.TP
.B \-\-modules
Equivalent to \fB\-\-debug=\fP\fI(LOADMASK)\fP.
.TP
.B \-\-algorithms
Equivalent to \fB\-\-debug=\fP\fI(ALGOMASK | ESTIMASK)\fP.
.TP
.B \-\-performance
Equivalent to \fB\-\-debug=\fP\fI(DEADBEEFMASK)\fP.
.TP
.B \-\-optimizers
Equivalent to \fB\-\-debug=\fP\fI(OPTMASK)\fP.
.TP
.B \-\-forcemito
Equivalent to \fB\-\-debug=\fP\fI(FORCEMITOMASK | NOSYNCMASK)\fP.
.SH MSERVER5 PARAMETERS
.I Mserver5
instructs the GDK kernel through the MAL (MonetDB Assembler Language)
language.
MonetDB 5 contains an extensive optimiser framework to transform MAL
plans into more optimal or functional (e.g.  distributed) plans.
These parameters control behaviour on the MAL level.
.TP
.B monet_vault_key
The authorisation tables inside
.I mserver5
can be encrypted with a key, such that reading the BATs does not
directly disclose any credentials.
The
.B monet_vault_key
setting points to a file that stores a secret key to unlock the
password vault.
It can contain anything.
The file is read up to the first null-byte ('\0'), hence it can be
padded to any length with trailing null-bytes to obfuscate the key
length.
Generating a key can be done for example by using a tool such as
.I pwgen
and adding a few of the passwords generated.
Make sure not to choose a too small key.
Note that on absence of a vault key file, some default key is used to
encrypt the authorisation tables.
Changing this setting (effectively changing the key) for an existing
database makes that database unusable as noone is any longer able to
login.
If you use
.IR monetdbd (1),
a per-database vault key is set.
.TP
.B max_clients
Controls how many client slots are allocated for clients to connect.
This settings limits the maximum number of connected clients at the
same time.
Note that MonetDB is not designed to handle massive amounts of
connected clients.
The funnel capability from
.IR monetdbd (1)
might be a more suitable solution for such workloads.

Default
.BR 64 .
.SH SQL PARAMETERS
The SQL component of MonetDB 5 runs on top of the MAL environment.
It has its own SQL-level specific settings.
.TP
.B sql_debug
Enable debugging using a mask.
This option should normally be disabled (0).
Default:
.BR 0 .
.TP
.B sql_optimizer
The default SQL optimizer pipeline can be set per server.
See the optpipe setting in
.IR monetdb (1)
when using monetdbd.
During SQL initialization, the optimizer pipeline is checked against
the dependency information maintained in the optimizer library to
ensure there are no conflicts and at least the pre-requisite
optimizers are used.
The setting of sql_optimizer can be either the list of optimizers to
run.
Default:
.BR default_pipe .
.RS
.P
The following are possible pipes to use:
.TP
.B minimal_pipe
The minimal pipeline necessary by the server to operate correctly.
.\" this documentation must be kept in sync with the respective code in monetdb5/optimizer/opt_pipes.c
minimal_pipe=inline,remap,deadcode,multiplex,generator,profiler,candidates,garbageCollector
.TP
.B default_pipe
The default pipeline contains the mitosis-mergetable-reorder
optimizers, aimed at large tables and improved access locality.
.\" this documentation must be kept in sync with the respective code in monetdb5/optimizer/opt_pipes.c
default_pipe=inline,remap,costModel,coercions,aliases,evaluate,emptybind,pushselect,aliases,mitosis,mergetable,deadcode,aliases,constants,commonTerms,projectionpath,deadcode,reorder,matpack,dataflow,querylog,multiplex,generator,profiler,candidates,postfix,deadcode,wlc,garbageCollector
.TP
.B no_mitosis_pipe
The no_mitosis pipeline is identical to the default pipeline, except
that optimizer mitosis is omitted.
It is used mainly to make some tests work deterministically, and to
check/debug whether "unexpected" problems are related to mitosis
(and/or mergetable).
.\" this documentation must be kept in sync with the respective code in monetdb5/optimizer/opt_pipes.c
no_mitosis_pipe=inline,remap,costModel,coercions,aliases,evaluate,emptybind,pushselect,aliases,mergetable,deadcode,aliases,constants,commonTerms,projectionpath,deadcode,reorder,matpack,dataflow,querylog,multiplex,generator,profiler,candidates,postfix,deadcode,wlc,garbageCollector
.TP
.B sequential_pipe
The sequential pipeline is identical to the default pipeline, except
that optimizers mitosis & dataflow are omitted.
It is use mainly to make some tests work deterministically, i.e.,
avoid ambigious output, by avoiding parallelism.
.\" this documentation must be kept in sync with the respective code in monetdb5/optimizer/opt_pipes.c
sequential_pipe=inline,remap,costModel,coercions,aliases,evaluate,emptybind,pushselect,aliases,mergetable,deadcode,aliases,constants,commonTerms,projectionpath,deadcode,reorder,matpack,querylog,multiplex,generator,profiler,candidates,postfix,deadcode,wlc,garbageCollector
.RE
.TP
.B embedded_py
Enable embedded Python.  This means Python code can be called from
SQL.
The value is
.B true
or
.B 2
for embedded Python 2, or
.B 3
for embedded Python 3.  Note that by enabling embedded Python, users
of the server are allowed to execute arbitrary Python code, and are
therefore able to read and modify all data that the server process has
access to.  Also note that it is not possible to enable both Python 2
and 3 at the same time.
.TP
.B embedded_r=true
Enable embedded R.  This means R code can be called from SQL.  Note
that by enabling embedded R, users of the server are allowed to
execute arbitrary R code, and are therefore able to read and modify
all data that the server process has access to.
.TP
.B embedded_c=true
Enable embedded C.  This means C code can be called from SQL.  The C
code will first be compiled and then executed.  This means a C
compiler must be available.  Note also that by enabling embedded C,
users of the server are allowed to execute arbitrary C code, and are

sql/ChangeLog
# ChangeLog file for sql
# This file is updated with Maddlog

* Wed Feb 18 2015 Sjoerd Mullender <sjoerd@acm.org>
- Removed support for the mseed library.

* Tue Feb 17 2015 Sjoerd Mullender <sjoerd@acm.org>
- Removed DataCell.  It was experimental code that was never enabled.



tools/mserver/mserver5.c
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2020 MonetDB B.V.
 */

#include "monetdb_config.h"
#include <locale.h>
#include "monet_options.h"
#include "mal.h"
#include "mal_session.h"
#include "mal_import.h"
#include "mal_client.h"
#include "mal_function.h"
#include "monet_version.h"
#include "mal_authorize.h"
#include "msabaoth.h"
#include "mutils.h"
#include <signal.h>

#ifdef NATIVE_WIN32
#include <crtdbg.h>
#include "monet_getopt.h"
#define getcwd _getcwd
#else
#include "getopt.h"
#endif

/* NEEDED? */
#if defined(_MSC_VER) && defined(__cplusplus)
#include <eh.h>
void
mserver_abort()
{
	fprintf(stderr, "\n! mserver_abort() was called by terminate(). !\n");
	fflush(stderr);
	exit(0);
}
#endif

#ifdef _MSC_VER
static void
mserver_invalid_parameter_handler(
	const wchar_t *expression,
	const wchar_t *function,
	const wchar_t *file,
	unsigned int line,
	uintptr_t reserved)
{
	(void) expression;
	(void) function;
	(void) file;
	(void) line;
	(void) reserved;
	/* the essential bit of this function is that it returns:
	 * we don't want the server to quit when a function is called
	 * with an invalid parameter */
}
#endif

static _Noreturn void usage(char *prog, int xit);

static void
usage(char *prog, int xit)
{
	fprintf(stderr, "Usage: %s [options]\n", prog);
	fprintf(stderr, "    --dbpath=<directory>      Specify database location\n");
	fprintf(stderr, "    --dbextra=<directory>     Directory for transient BATs\n");
	fprintf(stderr, "    --in-memory               Run database in-memory only\n");
	fprintf(stderr, "    --config=<config_file>    Use config_file to read options from\n");
	fprintf(stderr, "    --single-user             Allow only one user at a time\n");
	fprintf(stderr, "    --readonly                Safeguard database\n");
	fprintf(stderr, "    --set <option>=<value>    Set configuration option\n");
	fprintf(stderr, "    --help                    Print this list of options\n");
	fprintf(stderr, "    --version                 Print version and compile time info\n");
	fprintf(stderr, "    --verbose[=value]         Set or increase verbosity level\n");

	fprintf(stderr, "The debug, testing & trace options:\n");
	fprintf(stderr, "     --threads\n");
	fprintf(stderr, "     --memory\n");
	fprintf(stderr, "     --io\n");
	fprintf(stderr, "     --heaps\n");
	fprintf(stderr, "     --properties\n");
	fprintf(stderr, "     --transactions\n");
	fprintf(stderr, "     --modules\n");
	fprintf(stderr, "     --algorithms\n");
	fprintf(stderr, "     --performance\n");
	fprintf(stderr, "     --optimizers\n");
	fprintf(stderr, "     --forcemito\n");
	fprintf(stderr, "     --debug=<bitmask>\n");

	exit(xit);
}

/*
 * Collect some global system properties to relate performance results later
 */
static void
monet_hello(void)
{
	double sz_mem_h;
	char  *qc = " kMGTPE";
	int qi = 0;

	printf("# MonetDB 5 server v%s", GDKversion());
	{
#ifdef MONETDB_RELEASE
		printf(" (%s)", MONETDB_RELEASE);
#else
		const char *rev = mercurial_revision();
		if (strcmp(rev, "Unknown") != 0)
			printf(" (hg id: %s)", rev);
#endif
	}
#ifndef MONETDB_RELEASE
	printf("\n# This is an unreleased version");
#endif
	printf("\n# Serving database '%s', using %d thread%s\n",
			GDKgetenv("gdk_dbname"),
			GDKnr_threads, (GDKnr_threads != 1) ? "s" : "");
	printf("# Compiled for %s/%zubit%s\n",
			HOST, sizeof(ptr) * 8,
#ifdef HAVE_HGE
			" with 128bit integers"
#else
			""
#endif
			);
	sz_mem_h = (double) MT_npages() * MT_pagesize();
	while (sz_mem_h >= 1000.0 && qi < 6) {
		sz_mem_h /= 1024.0;
		qi++;
	}
	printf("# Found %.3f %ciB available main-memory",
			sz_mem_h, qc[qi]);
	sz_mem_h = (double) GDK_mem_maxsize;
	qi = 0;
	while (sz_mem_h >= 1000.0 && qi < 6) {
		sz_mem_h /= 1024.0;
		qi++;
	}
	printf(" of which we use %.3f %ciB\n",
			sz_mem_h, qc[qi]);
#ifdef MONET_GLOBAL_DEBUG
	printf("# Database path:%s\n", GDKgetenv("gdk_dbpath"));
	printf("# Module path:%s\n", GDKgetenv("monet_mod_path"));
#endif
	printf("# Copyright (c) 1993 - July 2008 CWI.\n");
	printf("# Copyright (c) August 2008 - 2020 MonetDB B.V., all rights reserved\n");
	printf("# Visit https://www.monetdb.org/ for further information\n");

	// The properties shipped through the performance profiler
	(void) snprintf(monet_characteristics, sizeof(monet_characteristics),
			"{\n"
			"\"version\":\"%s\",\n"
			"\"release\":\"%s\",\n"
			"\"host\":\"%s\",\n"
			"\"threads\":\"%d\",\n"
			"\"memory\":\"%.3f %cB\",\n"
			"\"oid\":\"%zu\",\n"
			"\"packages\":["
#ifdef HAVE_HGE
			"\"huge\""
#endif
			"]\n}",
			GDKversion(),
#ifdef MONETDB_RELEASE
			MONETDB_RELEASE,
#else
			"unreleased",
#endif
			HOST, GDKnr_threads,
			sz_mem_h, qc[qi], sizeof(oid) * 8);
}

static str
absolute_path(str s)
{
	if (!MT_path_absolute(s)) {
		str ret = (str) GDKmalloc(strlen(s) + strlen(monet_cwd) + 2);

		if (ret)
			sprintf(ret, "%s%c%s", monet_cwd, DIR_SEP, s);
		return ret;
	}
	return GDKstrdup(s);
}

#define BSIZE 8192

static int
monet_init(opt *set, int setlen)
{
	/* determine Monet's kernel settings */
	if (GDKinit(set, setlen) != GDK_SUCCEED)
		return 0;
#ifndef NATIVE_WIN32
	setsid();
#endif
	monet_hello();
	return 1;
}

static void emergencyBreakpoint(void)
{
	/* just a handle to break after system initialization for GDB */
}

static volatile sig_atomic_t interrupted = 0;

#ifdef _MSC_VER
static BOOL WINAPI
winhandler(DWORD type)
{
	(void) type;
	interrupted = 1;
	return TRUE;
}
#else
static void
handler(int sig)
{
	(void) sig;
	interrupted = 1;
}
#endif

int
main(int argc, char **av)
{
	char *prog = *av;
	opt *set = NULL;
	int grpdebug = 0, debug = 0, setlen = 0;
	str err = MAL_SUCCEED;
	char prmodpath[FILENAME_MAX];
	const char *modpath = NULL;
	char *binpath = NULL;
	char *dbpath = NULL;
	char *dbextra = NULL;
	int verbosity = 0;
	bool inmemory = false;
	static struct option long_options[] = {
		{ "config", required_argument, NULL, 'c' },
		{ "dbpath", required_argument, NULL, 0 },
		{ "dbextra", required_argument, NULL, 0 },
		{ "debug", optional_argument, NULL, 'd' },
		{ "help", no_argument, NULL, '?' },
		{ "version", no_argument, NULL, 0 },
		{ "verbose", optional_argument, NULL, 'v' },
		{ "readonly", no_argument, NULL, 'r' },
		{ "single-user", no_argument, NULL, 0 },
		{ "set", required_argument, NULL, 's' },
		{ "threads", no_argument, NULL, 0 },
		{ "memory", no_argument, NULL, 0 },
		{ "properties", no_argument, NULL, 0 },
		{ "io", no_argument, NULL, 0 },
		{ "transactions", no_argument, NULL, 0 },
		{ "modules", no_argument, NULL, 0 },
		{ "algorithms", no_argument, NULL, 0 },
		{ "optimizers", no_argument, NULL, 0 },
		{ "performance", no_argument, NULL, 0 },
		{ "forcemito", no_argument, NULL, 0 },
		{ "heaps", no_argument, NULL, 0 },
		{ "in-memory", no_argument, NULL, 0 },
		{ NULL, 0, NULL, 0 }
	};

#if defined(_MSC_VER) && defined(__cplusplus)
	set_terminate(mserver_abort);
#endif
#ifdef _MSC_VER
	_CrtSetReportMode(_CRT_ERROR, 0);
	_CrtSetReportMode(_CRT_ASSERT, 0);
	_set_invalid_parameter_handler(mserver_invalid_parameter_handler);
#ifdef _TWO_DIGIT_EXPONENT
	_set_output_format(_TWO_DIGIT_EXPONENT);
#endif
#endif
	if (setlocale(LC_CTYPE, "") == NULL) {
		fprintf(stderr, "cannot set locale\n");
		exit(1);
	}

	if (getcwd(monet_cwd, FILENAME_MAX - 1) == NULL) {
		perror("pwd");
		fprintf(stderr,"monet_init: could not determine current directory\n");
		exit(-1);
	}

	/* retrieve binpath early (before monet_init) because some
	 * implementations require the working directory when the binary was
	 * called */
	binpath = get_bin_path();

	if (!(setlen = mo_builtin_settings(&set)))
		usage(prog, -1);

	for (;;) {
		int option_index = 0;

		int c = getopt_long(argc, av, "c:d::rs:t::v::?",
				long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 0:
			if (strcmp(long_options[option_index].name, "in-memory") == 0) {
				inmemory = true;
				break;
			}
			if (strcmp(long_options[option_index].name, "dbpath") == 0) {
				size_t optarglen = strlen(optarg);
				/* remove trailing directory separator */
				while (optarglen > 0 &&
				       (optarg[optarglen - 1] == '/' ||
					optarg[optarglen - 1] == '\\'))
					optarg[--optarglen] = '\0';
				dbpath = absolute_path(optarg);
				if( dbpath == NULL)
					fprintf(stderr, "#error: can not allocate memory for dbpath\n");
				else
					setlen = mo_add_option(&set, setlen, opt_cmdline, "gdk_dbpath", dbpath);
				break;
			}
			if (strcmp(long_options[option_index].name, "dbextra") == 0) {
				if (dbextra)
					fprintf(stderr, "#warning: ignoring multiple --dbextra arguments\n");
				else
					dbextra = optarg;
				break;
			}
			if (strcmp(long_options[option_index].name, "single-user") == 0) {
				setlen = mo_add_option(&set, setlen, opt_cmdline, "gdk_single_user", "yes");
				break;
			}
			if (strcmp(long_options[option_index].name, "version") == 0) {
				monet_version();
				exit(0);
			}
			/* debugging options */
			if (strcmp(long_options[option_index].name, "properties") == 0) {
				grpdebug |= GRPproperties;
				break;
			}
			if (strcmp(long_options[option_index].name, "algorithms") == 0) {
				grpdebug |= GRPalgorithms;
				break;
			}
			if (strcmp(long_options[option_index].name, "optimizers") == 0) {
				grpdebug |= GRPoptimizers;
				break;
			}
			if (strcmp(long_options[option_index].name, "forcemito") == 0) {
				grpdebug |= GRPforcemito;
				break;
			}
			if (strcmp(long_options[option_index].name, "performance") == 0) {
				grpdebug |= GRPperformance;
				break;
			}
			if (strcmp(long_options[option_index].name, "io") == 0) {
				grpdebug |= GRPio;
				break;
			}
			if (strcmp(long_options[option_index].name, "memory") == 0) {
				grpdebug |= GRPmemory;
				break;
			}
			if (strcmp(long_options[option_index].name, "modules") == 0) {
				grpdebug |= GRPmodules;
				break;
			}
			if (strcmp(long_options[option_index].name, "transactions") == 0) {
				grpdebug |= GRPtransactions;
				break;
			}
			if (strcmp(long_options[option_index].name, "threads") == 0) {
				grpdebug |= GRPthreads;
				break;
			}
			if (strcmp(long_options[option_index].name, "heaps") == 0) {
				grpdebug |= GRPheaps;
				break;
			}
			usage(prog, -1);
			/* not reached */
		case 'c':
			/* coverity[var_deref_model] */
			setlen = mo_add_option(&set, setlen, opt_cmdline, "config", optarg);
			break;
		case 'd':
			if (optarg) {
				char *endarg;
				debug |= strtol(optarg, &endarg, 10);
				if (*endarg != '\0') {
					fprintf(stderr, "ERROR: wrong format for --debug=%s\n",
							optarg);
					usage(prog, -1);
				}
			} else {
				debug |= 1;
			}
			break;
		case 'r':
			setlen = mo_add_option(&set, setlen, opt_cmdline, "gdk_readonly", "yes");
			break;
		case 's': {
			/* should add option to a list */
			/* coverity[var_deref_model] */
			char *tmp = strchr(optarg, '=');

			if (tmp) {
				*tmp = '\0';
				setlen = mo_add_option(&set, setlen, opt_cmdline, optarg, tmp + 1);
			} else
				fprintf(stderr, "ERROR: wrong format %s\n", optarg);
			}
			break;
		case 'v':
			if (optarg) {
				char *endarg;
				verbosity = (int) strtol(optarg, &endarg, 10);
				if (*endarg != '\0') {
					fprintf(stderr, "ERROR: wrong format for --verbose=%s\n",
							optarg);
					usage(prog, -1);
				}
			} else {
				verbosity++;
			}
			break;
		case '?':
			/* a bit of a hack: look at the option that the
			   current `c' is based on and see if we recognize
			   it: if -? or --help, exit with 0, else with -1 */
			usage(prog, strcmp(av[optind - 1], "-?") == 0 || strcmp(av[optind - 1], "--help") == 0 ? 0 : -1);
		default:
			fprintf(stderr, "ERROR: getopt returned character "
				"code '%c' 0%o\n", c, (uint8_t) c);
			usage(prog, -1);
		}
	}

	if (optind < argc)
		usage(prog, -1);

	if (!(setlen = mo_system_config(&set, setlen)))
		usage(prog, -1);

	GDKsetdebug(debug | grpdebug);  /* add the algorithm tracers */
	if (debug)
		mo_print_options(set, setlen);
	GDKsetverbose(verbosity);

	if (dbpath && inmemory) {
		fprintf(stderr, "!ERROR: both dbpath and in-memory must not be set at the same time\n");
		exit(1);
	}

	if (!dbpath) {
		dbpath = absolute_path(mo_find_option(set, setlen, "gdk_dbpath"));
		if (!dbpath) {
			fprintf(stderr, "!ERROR: cannot allocate memory for database directory \n");
			exit(1);
		}
	}
	if (inmemory) {
		if (BBPaddfarm(NULL, (1 << PERSISTENT) | (1 << TRANSIENT)) != GDK_SUCCEED) {
			fprintf(stderr, "!ERROR: cannot add in-memory farm\n");
			exit(1);
		}
	} else {
		if (BBPaddfarm(dbpath, 1 << PERSISTENT) != GDK_SUCCEED ||
		    BBPaddfarm(dbextra ? dbextra : dbpath, 1 << TRANSIENT) != GDK_SUCCEED) {
			fprintf(stderr, "!ERROR: cannot add farm\n");
			exit(1);
		}
		if (GDKcreatedir(dbpath) != GDK_SUCCEED) {
			fprintf(stderr, "!ERROR: cannot create directory for %s\n", dbpath);
			exit(1);
		}
	}
	GDKfree(dbpath);
	if (monet_init(set, setlen) == 0) {
		mo_free_options(set, setlen);
		if (GDKerrbuf && *GDKerrbuf)
			fprintf(stderr, "%s\n", GDKerrbuf);
		exit(1);
	}
	mo_free_options(set, setlen);

	if (GDKsetenv("monet_version", GDKversion()) != GDK_SUCCEED ||
	    GDKsetenv("monet_release",
#ifdef MONETDB_RELEASE
		      MONETDB_RELEASE
#else
		      "unreleased"
#endif
		    ) != GDK_SUCCEED) {
		fprintf(stderr, "!ERROR: GDKsetenv failed\n");
		exit(1);
	}

	if ((modpath = GDKgetenv("monet_mod_path")) == NULL) {
		/* start probing based on some heuristics given the binary
		 * location:
		 * bin/mserver5 -> ../
		 * libX/monetdb5/lib/
		 * probe libX = lib, lib32, lib64, lib/64 */
		size_t pref;
		/* "remove" common prefix of configured BIN and LIB
		 * directories from LIBDIR */
		for (pref = 0; LIBDIR[pref] != 0 && BINDIR[pref] == LIBDIR[pref]; pref++)
			;
		const char *libdirs[] = {
			&LIBDIR[pref],
			"lib",
			"lib64",
			"lib/64",
			"lib32",
			NULL,
		};
		struct stat sb;
		if (binpath != NULL) {
			char *p = strrchr(binpath, DIR_SEP);
			if (p != NULL)
				*p = '\0';
			p = strrchr(binpath, DIR_SEP);
			if (p != NULL) {
				*p = '\0';
				for (int i = 0; libdirs[i] != NULL; i++) {
					int len = snprintf(prmodpath, sizeof(prmodpath), "%s%c%s%cmonetdb5",
									   binpath, DIR_SEP, libdirs[i], DIR_SEP);
					if (len == -1 || len >= FILENAME_MAX)
						continue;
					if (stat(prmodpath, &sb) == 0) {
						modpath = prmodpath;
						break;
					}
				}
			} else {
				printf("#warning: unusable binary location, "
					   "please use --set monet_mod_path=/path/to/... to "
					   "allow finding modules\n");
				fflush(NULL);
			}
		} else {
			printf("#warning: unable to determine binary location, "
				   "please use --set monet_mod_path=/path/to/... to "
				   "allow finding modules\n");
			fflush(NULL);
		}
		if (modpath != NULL &&
		    GDKsetenv("monet_mod_path", modpath) != GDK_SUCCEED) {
			fprintf(stderr, "!ERROR: GDKsetenv failed\n");
			exit(1);
		}
	}

	if (!GDKinmemory()) {
		/* configure sabaoth to use the right dbpath and active database */
		msab_dbpathinit(GDKgetenv("gdk_dbpath"));
		/* wipe out all cruft, if left over */
		if ((err = msab_wildRetreat()) != NULL) {
			/* just swallow the error */
			free(err);
		}
		/* From this point, the server should exit cleanly.  Discussion:
		 * even earlier?  Sabaoth here registers the server is starting up. */
		if ((err = msab_registerStarting()) != NULL) {
			/* throw the error at the user, but don't die */
			fprintf(stderr, "!%s\n", err);
			free(err);
		}
	}

#ifndef NATIVE_WIN32
	{
		struct sigaction sa;

		(void) sigemptyset(&sa.sa_mask);
		sa.sa_flags = 0;
		sa.sa_handler = handler;
		if (sigaction(SIGINT, &sa, NULL) == -1 ||
		    sigaction(SIGQUIT, &sa, NULL) == -1 ||
		    sigaction(SIGTERM, &sa, NULL) == -1) {
			fprintf(stderr, "!unable to create signal handlers\n");
		}
	}
#else
#ifdef _MSC_VER
	if (!SetConsoleCtrlHandler(winhandler, TRUE))
		fprintf(stderr, "!unable to create console control handler\n");
#else
	if(signal(SIGINT, handler) == SIG_ERR)
		fprintf(stderr, "!unable to create signal handlers\n");
#ifdef SIGQUIT
	if(signal(SIGQUIT, handler) == SIG_ERR)
		fprintf(stderr, "!unable to create signal handlers\n");
#endif
	if(signal(SIGTERM, handler) == SIG_ERR)
		fprintf(stderr, "!unable to create signal handlers\n");
#endif
#endif

	if (!GDKinmemory()) {
		str lang = "mal";
		/* we inited mal before, so publish its existence */
		if ((err = msab_marchScenario(lang)) != NULL) {
			/* throw the error at the user, but don't die */
			fprintf(stderr, "!%s\n", err);
			free(err);
		}
	}

	{
		/* unlock the vault, first see if we can find the file which
		 * holds the secret */
		char secret[1024];
		char *secretp = secret;
		FILE *secretf;
		size_t len;

		if (GDKinmemory() || GDKgetenv("monet_vault_key") == NULL) {
			/* use a default (hard coded, non safe) key */
			snprintf(secret, sizeof(secret), "%s", "Xas632jsi2whjds8");
		} else {
			if ((secretf = fopen(GDKgetenv("monet_vault_key"), "r")) == NULL) {
				fprintf(stderr,
					"unable to open vault_key_file %s: %s\n",
					GDKgetenv("monet_vault_key"), strerror(errno));
				/* don't show this as a crash */
				msab_registerStop();
				exit(1);
			}
			len = fread(secret, 1, sizeof(secret), secretf);
			secret[len] = '\0';
			len = strlen(secret); /* secret can contain null-bytes */
			if (len == 0) {
				fprintf(stderr, "vault key has zero-length!\n");
				/* don't show this as a crash */
				msab_registerStop();
				exit(1);
			} else if (len < 5) {
				fprintf(stderr, "#warning: your vault key is too short "
								"(%zu), enlarge your vault key!\n", len);
			}
			fclose(secretf);
		}
		if ((err = AUTHunlockVault(secretp)) != MAL_SUCCEED) {
			/* don't show this as a crash */
			if (!GDKinmemory())
				msab_registerStop();
			fprintf(stderr, "%s\n", err);
			freeException(err);
			exit(1);
		}
	}
	/* make sure the authorisation BATs are loaded */
	if ((err = AUTHinitTables(NULL)) != MAL_SUCCEED) {
		/* don't show this as a crash */
		if (!GDKinmemory())
			msab_registerStop();
		fprintf(stderr, "%s\n", err);
		freeException(err);
		exit(1);
	}
	if (mal_init()) {
		/* don't show this as a crash */
		if (!GDKinmemory())
			msab_registerStop();
		return 0;
	}

	emergencyBreakpoint();

	if (!GDKinmemory() && (err = msab_registerStarted()) != NULL) {
		/* throw the error at the user, but don't die */
		fprintf(stderr, "!%s\n", err);
		free(err);
	}

#ifdef _MSC_VER
	printf("# MonetDB server is started. To stop server press Ctrl-C.\n");
#endif

	/* why busy wait ? */
	while (!interrupted && !GDKexiting()) {
		MT_sleep_ms(100);
	}

	/* mal_exit calls exit, so statements after this call will
	 * never get reached */
	mal_exit(0);

	return 0;
}
